New docs area for URL semantics, relocate some HTTP pages here (#35202)
mdn/content: files/en-us/web/uri/index.md

---
title: URIs
slug: Web/URI
page-type: guide
spec-urls: https://www.rfc-editor.org/rfc/rfc3986.html
---

{{QuickLinksWithSubpages("/en-US/docs/Web/URI")}}

> **Uniform Resource Identifiers (URI)** are used to identify "resources" on the web. They are commonly used as targets of [HTTP](/en-US/docs/Web/HTTP) requests, in which case the URI represents a location for a physical resource, such as a document, a photo, binary data, etc. URLs can also be used to trigger behaviors other than fetching a resource, including opening the email client, sending text messages, or executing JavaScript, when used in other places such as the [`href`](/en-US/docs/Web/HTML/Element/a#href) of an HTML `<a>` link.

Reviewer: One case where we can do `URL` -> `URI`:

```suggestion
**Uniform Resource Identifiers (URI)** are used to identify "resources" on the web. They are commonly used as targets of [HTTP](/en-US/docs/Web/HTTP) requests, in which case the URI represents a location for a physical resource, such as a document, a photo, binary data, etc. URIs can also be used to trigger behaviors other than fetching a resource, including opening the email client, sending text messages, or executing JavaScript, when used in other places such as the [`href`](/en-US/docs/Web/HTML/Element/a#href) of an HTML `<a>` link.
```

## URLs and URNs

### URLs

The most common form of URI is the Uniform Resource Locator ({{Glossary("URL")}}), which is known as the _web address_.

```url
https://developer.mozilla.org
https://developer.mozilla.org/en-US/docs/Learn/
https://developer.mozilla.org/en-US/search?q=URL
```

Any of those URLs can be typed into your browser's address bar to tell it to load the associated page (resource).

> A URL is composed of different parts, some mandatory and others optional. A more complex example might look like this:

Reviewer: Do you want to mention that the parts of a URL are described in detail below?

Author: I think this is fine as-is. The details below aren't directly dependent on this example anyway.

---
title: URIs
slug: Web/URI
page-type: guide
spec-urls: https://httpwg.org/specs/rfc9110.html#uri
---

> {{QuickLinksWithSubpages("/en-US/docs/Web/URI")}}

Reviewer 2: This sidebar is not being generated in preview. This is a known thing or some kind of error in the declaration?

Author: It's because there's actually no immediate "subpage" of this landing page 😅 Should get better after we write pages for "scheme", "authority", etc. Or we can just make a new sidebar in Yari.

Reviewer 2: Not working for me locally. I suspect it can't cope with the fact that the /authority sub folder has no index. If you create such a topic, it would be good to link that from the authority of the URL top level page.

Author: I'm aware there aren't index pages for the URI parts. I would write them in a second PR and this one is focused on the move and the creation of this subtree so hopefully there's less to review.

Reviewer 2: IMO there needs to be a complete sidebar for this to go in.

Reviewer 2: Note, appreciate that this is a pain. I'm happy to do the reviews for the new pages in a timely manner.

> **URIs** are used to identify "resources" on the web. URIs are commonly used as targets of [HTTP](/en-US/docs/Web/HTTP) requests, in which case the URI represents a location for a physical resource, such as a document, a photo, binary data, etc. URLs can also be used to trigger behaviors other than fetching a resource, including opening the email client, sending text messages, or executing JavaScript, when used in other places such as the [`href`](/en-US/docs/Web/HTML/Element/a#href) of an HTML `<a>` link.

Reviewer 2: ```suggestion
**Uniform Resource Identifiers (URI)** are used to identify "resources" on the web. URIs are commonly used as targets of [HTTP](/en-US/docs/Web/HTTP) requests, in which case the URI represents a location for a physical resource, such as a document, a photo, binary data, etc. URLs can also be used to trigger behaviors other than fetching a resource, including opening the email client, sending text messages, or executing JavaScript, when used in other places such as the [`href`](/en-US/docs/Web/HTML/Element/a#href) of an HTML `<a>` link.
```

## URLs and URNs

### URLs

The most common form of URI is the Uniform Resource Locator ({{Glossary("URL")}}), which is known as the _web address_.

```url
https://developer.mozilla.org
https://developer.mozilla.org/en-US/docs/Learn/
https://developer.mozilla.org/en-US/search?q=URL
```

Any of those URLs can be typed into your browser's address bar to tell it to load the associated page (resource).

A URL is composed of different parts, some mandatory and others optional. A more complex example might look like this:

```url
http://www.example.com:80/path/to/myfile.html?key1=value1&key2=value2#SomewhereInTheDocument
```

### URNs

> A Uniform Resource Name (URN) is a URI that identifies a resource by name in a particular namespace.

Reviewer 2: Perhaps not for now, but it might be worth expanding out what this namespace thing "means" i.e it provides some predefined context to the address that allows it to be expanded to a particular address on the Internet.

| ssh                              | Secure shell                                                         |
| tel                              | telephone                                                            |
| urn                              | Uniform Resource Names                                               |
| view-source                      | Source code of the resource                                          |
| ws/wss                           | [WebSocket connections (Secure)](/en-US/docs/Web/API/WebSockets_API) |

When using URLs in {{Glossary("HTML")}} content, you should generally only use a few of these URL schemes. When referring to subresources — that is, files that are being loaded as part of a larger document — you should only use the HTTP and HTTPS schemes. Increasingly, browsers are removing support for using FTP to load subresources, for security reasons.

FTP is still acceptable at the top level (such as typed directly into the browser's URL bar, or the target of a link), although some browsers may delegate loading FTP content to another application.

### Authority

![Domain Name](mdn-url-domain@x2.png)

> `www.example.com` is the domain name or authority that governs the namespace. It indicates which Web server is being requested. Alternatively, it is possible to directly use an {{Glossary("IP address")}}, but because it is less convenient, it is not often used on the Web.

Reviewer 2: ```suggestion
`www.example.com` is the domain name that governs the namespace, and is referred to in the corresponding specification as the "authority". It indicates which Web server is being requested. Alternatively, it is possible to directly use an {{Glossary("IP address")}}, but because it is less convenient, it is not often used on the Web.
```

### Port

![Port](mdn-url-port@x2.png)

> `:80` is the port in this instance. It indicates the technical "gate" used to access the resources on the web server. It is usually omitted if the web server uses the standard ports of the HTTP protocol (80 for HTTP and 443 for HTTPS) to grant access to its resources. Otherwise, it is mandatory.

Reviewer 2: "In this instance"? Assuming you mean that this can be something else in other contexts I would say something like:

```suggestion
In HTTP and HTTP addresses, `:80` is the "port", which indicates the technical "gate" used to access the resources on the web server. It is usually omitted if the web server uses the standard ports of the HTTP protocol (80 for HTTP and 443 for HTTPS) to grant access to its resources. Otherwise, it is mandatory.

In other contexts the port can be ...
```

### Path

![Path to the file](mdn-url-path@x2.png)

> `/path/to/myfile.html` is the path to the resource on the Web server. In the early days of the Web, a path like this represented a physical file location on the Web server. Nowadays, it is mostly an abstraction handled by Web servers without any physical reality.

Reviewer 2: "is the path" is perhaps better as "is the location" to capture your following point.
```suggestion
`/path/to/myfile.html` is the location of the resource on the Web server. In the early days of the Web, this was an actual directory path to a physical location on the Web server. Nowadays, Web servers usually abstract this to an arbitrary location.
```

### Query

![Parameters](mdn-url-parameters@x2.png)

> `?key1=value1&key2=value2` are extra parameters provided to the Web server. Those parameters are a list of key/value pairs separated with the `&` symbol. The Web server can use those parameters to do extra stuff before returning the resource to the user. Each Web server has its own rules regarding parameters, and the only reliable way to know how a specific Web server is handling parameters is by asking the Web server owner.

Reviewer 2: ```suggestion
`?key1=value1&key2=value2` are extra parameters provided to the Web server. The parameters are a list of key/value pairs prefixed by the `?` symbol, and separated with the `&` symbol, that can be used to provide additional context about the resource being requested. Each Web server has its own rules regarding parameters, and the only reliable way to know how a specific Web server is handling parameters is by asking the owner of the server.
```

| view-source                      | Source code of the resource                                          |
| ws/wss                           | [WebSocket connections (Secure)](/en-US/docs/Web/API/WebSockets_API) |

When using URLs in {{Glossary("HTML")}} content, you should generally only use a few of these URL schemes. When referring to subresources — that is, files that are being loaded as part of a larger document — you should only use the HTTP and HTTPS schemes. Increasingly, browsers are removing support for using FTP to load subresources, for security reasons.

FTP is still acceptable at the top level (such as typed directly into the browser's URL bar, or the target of a link), although some browsers may delegate loading FTP content to another application.

### Host name

![Domain Name](mdn-url-domain@x2.png)

`www.example.com` is the _host name_, or domain name, indicating which Web server is being requested. Alternatively, it is possible to directly use an {{Glossary("IP address")}}, but because it is less convenient, it is not often used on the Web.

### Port

![Port](mdn-url-port@x2.png)

> `:80` is the _port_ or the URL, indicating the technical "gate" used to access the resources on the web server. It is usually omitted if the web server uses the standard ports of the HTTP protocol (80 for HTTP and 443 for HTTPS) to grant access to its resources. Otherwise, it is mandatory.

Reviewer 2: ```suggestion
`:80` is the _port_ of the URL, indicating the technical "gate" used to access the resources on the web server. It is usually omitted if the web server uses the standard ports of the HTTP protocol (80 for HTTP and 443 for HTTPS) to grant access to its resources. Otherwise, it is mandatory.
```

```url
http://www.example.com:80/path/to/myfile.html?key1=value1&key2=value2#SomewhereInTheDocument
```

### URNs

A Uniform Resource Name (URN) is a URI that identifies a resource by name in a particular namespace.

```url
urn:isbn:9780141036144
urn:ietf:rfc:7230
```

The two URNs correspond to

- the book Nineteen Eighty-Four by George Orwell,
- the IETF specification 7230, Hypertext Transfer Protocol (HTTP/1.1): Message Syntax and Routing.

## Syntax of Uniform Resource Identifiers (URIs)

We will break down the following URL into its components:

```url
http://www.example.com:80/path/to/myfile.html?key1=value1&key2=value2#SomewhereInTheDocument
```

### Scheme

![Protocol](mdn-url-protocol@x2.png)

`http://` is the [_scheme_](/en-US/docs/Web/URI/Schemes) of the URL, indicating which protocol the browser must use. Usually it is the HTTP protocol or its secured version, HTTPS. The Web requires one of these two, but browsers also know how to handle other protocols such as `mailto:` (to open a mail client) or `ftp:` to handle a file transfer, so don't be surprised if you see such protocols. Common schemes are:

> - [`blob`](/en-US/docs/Web/API/URL/createObjectURL_static)

Reviewer 2: These are all expanded out in the scheme docs, now we have one.  An option here is to just list a few of the important ones here and note that there are many others. I would do that, but leave to you to decide.

- `ftp`
  - : {{Glossary("FTP","File Transfer Protocol")}}
- `http/https`
  - : [Hyper text transfer protocol (Secure)](/en-US/docs/Glossary/HTTP)
- `javascript`
  - : URL-embedded JavaScript code
- `mailto`
  - : Electronic mail address
- [`resource`](/en-US/docs/Web/URI/Schemes/resource) {{Non-standard_inline}}
  - : Firefox and Firefox browser extensions to load resources internally
- `ssh`
  - : Secure shell
- `tel`
  - : telephone
- `urn`
  - : Uniform Resource Names
- `view-source`
  - : Source code of the resource
- `ws/wss`
  - : [WebSocket connections (Secure)](/en-US/docs/Web/API/WebSockets_API)

When using URLs in {{Glossary("HTML")}} content, you should generally only use a few of these URL schemes. When referring to subresources — that is, files that are being loaded as part of a larger document — you should only use the HTTP and HTTPS schemes. Increasingly, browsers are removing support for using FTP to load subresources, for security reasons.

FTP is still acceptable at the top level (such as typed directly into the browser's URL bar, or the target of a link), although some browsers may delegate loading FTP content to another application.

### Authority

> 

Reviewer 2: I would prefer the last sentence moved here or immediately below the image (I know this breaks the pattern). Also that this mentions the user section too. This is better because you immediately have the definition and the link to more detail.

So something like.

```suggestion

The URI [authority](/en-US/docs/Web/URI/Authority) is comprised of user credentials (optional - not shown), host name, and port.

```

> ![Domain Name](mdn-url-domain@x2.png)

Reviewer 2: Diagram does not include port as part of the authority.

Author: Diagram calls it "domain name", which is correct. Authority = domain name + port

Reviewer 2: True.

![Domain Name](mdn-url-domain@x2.png)

> `www.example.com` is the _host name_ of the URI, indicating which Web server is being requested. Here, we use a domain name; alternatively, it is possible to directly use an {{Glossary("IP address")}}, but because it is less convenient, it is not often used on the Web.

Reviewer 2: ```suggestion
`www.example.com` is the _host name_ of the URI, indicating which Web server is being requested. Here, we use a domain name. It is also possible to directly use an {{Glossary("IP address")}}, but because it is less convenient, it is rare to do so.
```

Author: ```suggestion
`www.example.com` is the _host name_ of the URI, indicating which Web server is being requested. Here, we use a domain name. It is also possible to directly use an {{Glossary("IP address")}}, but because it is less convenient, it is rare to do so, unless the server doesn't have a registered domain name.
```

FTP is still acceptable at the top level (such as typed directly into the browser's URL bar, or the target of a link), although some browsers may delegate loading FTP content to another application.

### Host name

![Domain Name](mdn-url-domain@x2.png)

> `www.example.com` is the _host name_, or domain name, indicating which Web server is being requested. Alternatively, it is possible to directly use an {{Glossary("IP address")}}, but because it is less convenient, it is not often used on the Web.

Reviewer 2: Good fix.
